chia/wallet/wallet_node.py

```diff
@@ -498,27 +498,30 @@ async def on_connect(self, peer: WSChiaConnection):
 
     async def perform_atomic_rollback(self, fork_height: int, cache: Optional[PeerRequestCache] = None):
         self.log.info(f"perform_atomic_rollback to {fork_height}")
-        async with self.wallet_state_manager.db_wrapper.lock:
+        # this is to start a write transaction
+        async with self.wallet_state_manager.db_wrapper.write_db():
             try:
-                await self.wallet_state_manager.db_wrapper.begin_transaction()
                 removed_wallet_ids = await self.wallet_state_manager.reorg_rollback(fork_height)
-                await self.wallet_state_manager.blockchain.set_finished_sync_up_to(fork_height, True, in_rollback=True)
+                await self.wallet_state_manager.blockchain.set_finished_sync_up_to(fork_height, in_rollback=True)
                 if cache is None:
                     self.rollback_request_caches(fork_height)
                 else:
                     cache.clear_after_height(fork_height)
-                await self.wallet_state_manager.db_wrapper.commit_transaction()
             except Exception as e:
                 tb = traceback.format_exc()
                 self.log.error(f"Exception while perform_atomic_rollback: {e} {tb}")
-                await self.wallet_state_manager.db_wrapper.rollback_transaction()
                 raise
             else:
                 await self.wallet_state_manager.blockchain.clean_block_records()
 
                 for wallet_id in removed_wallet_ids:
                     self.wallet_state_manager.wallets.pop(wallet_id)
 
+        # this has to be called *after* the transaction commits, otherwise it
+        # won't see the changes (since we spawn a new task to handle potential
+        # resends)
+        self._pending_tx_handler()
```

Comment:
this change is contained in the last commit. If fixes the issue described in this comment. Note taht `wallet_state_manager` does not control when its changes are committed to the database, since we're wrapping it in an outer transaction here.